gpiozero/pins/pi.py

```diff
@@ -0,0 +1,214 @@
+        # to ensure the shared-key for shared implementations get matched
+        # correctly, and is a bit of a hack for the pigpio bit-bang
+        # implementation which just wants the pin numbers too.
+        spi_args = {
+            key: pin.number if isinstance(pin, Pin) else pin
+            for key, pin in spi_args.items()
+            }
+        if shared:
+            return self.shared_spi_software_class(self, **spi_args)
+        else:
+            return self.spi_software_class(self, **spi_args)
+
+    def _extract_spi_args(self, **kwargs):
+        """
+        Given a set of keyword arguments, splits it into those relevant to SPI
+        implementations and all the rest. SPI arguments are augmented with
+        defaults and converted into the pin format (from the port/device
+        format) if necessary.
+
+        Returns a tuple of ``(spi_args, other_args)``.
+        """
+        pin_defaults = {
+            'clock_pin': 11,
+            'mosi_pin': 10,
+            'miso_pin': 9,
+            'select_pin': 8,
+            }
+        dev_defaults = {
+            'port': 0,
+            'device': 0,
+            }
+        spi_args = {
+            key: value for (key, value) in kwargs.items()
+            if key in pin_defaults or key in dev_defaults
+            }
+        kwargs = {
+            key: value for (key, value) in kwargs.items()
+            if key not in spi_args
+            }
+        if not spi_args:
+            spi_args = pin_defaults
+        elif set(spi_args) <= set(pin_defaults):
+            spi_args = {
+                key: self._to_gpio(spi_args.get(key, default))
+                for key, default in pin_defaults.items()
+                }
+        elif set(spi_args) <= set(dev_defaults):
+            spi_args = {
+                key: spi_args.get(key, default)
+                for key, default in dev_defaults.items()
+                }
+            if spi_args['port'] != 0:
+                raise SPIBadArgs('port 0 is the only valid SPI port')
+            if spi_args['device'] not in (0, 1):
+                raise SPIBadArgs('device must be 0 or 1')
+            spi_args = {
+                key: value if key != 'select_pin' else (8, 7)[spi_args['device']]
+                for key, value in pin_defaults.items()
+                }
+        else:
+            raise SPIBadArgs(
+                'you must either specify port and device, or clock_pin, '
+                'mosi_pin, miso_pin, and select_pin; combinations of the two '
+                'schemes (e.g. port and clock_pin) are not permitted')
+        return spi_args, kwargs
+
+
+class PiPin(Pin):
+    """
+    Abstract base class representing a multi-function GPIO pin attached to a
+    Raspberry Pi.
+    """
+    def __init__(self, factory, number):
+        super(PiPin, self).__init__()
+        try:
+            factory.pi_info.physical_pin('GPIO%d' % number)
```

Comment:
Given that there's **_so_** many occurrences of `'GPIO%d'`scattered throughout the code, would a `_get_str_name` type method be useful?